contents/scalability/scalable-development.md

```diff
@@ -0,0 +1,71 @@
+
+Here are some tips to get started with horizontal scaling:
+- Understand parallel computing and how to use it to speed up computation.
+- Learn how to utilize MapReduce on [Hadoop](https://www.tutorialspoint.com/hadoop/index.htm). The Hadoop ecosystem is so versatile and widely-used that there are companies offering Hadoop as a service, so you do not have to set up the server yourself
+- Host your server on [Google App Engine](https://cloud.google.com/appengine/docs). Despite being more expensive than Amazon Web Services and DigitalOcean, it offers very scalable infrastructure for your server.
+- Run a server with multiple nodes with MPI. [Open MPI](https://www.open-mpi.org/) is an open source message-passing library that can be used to send data between the nodes in your server. You might have to study [cloud computing infrastructure](http://whatiscloud.com/basic_concepts_and_terminology/it_resource) in order to make use of this.
+- Study various scalable architecture for servers ([resource](http://srinathsview.blogspot.sg/2011/10/list-of-known-scalable-architecture.html)).
+
+## But wait!
+
+Before you start tweaking your server to make it scalable, be sure to avoid the following pitfalls:
+- Premature or unnecessary optimization: a tiny performance improvement that can make your code hard to understand later on is not worth it.
+- Optimizing without profiling performance: optimizing the function that takes up 1% of the delay will not improve a lot.
+- Forget the product: if there is no product, then scaling your server to handle a million users is a moot point.
+
+Use the following tips to avoid those pitfalls:
+- Structure the code so that it will be easy to scale up. For instance if you know a portion of your code can be implemented with MapReduce, code it so that it follows the MapReduce structure.
+- Implement code that is easy to understand. If your "smart" tweak saves you half a millisecond makes your code hard to understand, then it probably is a premature optimization.
+- Always profile your program's performance first. Be sure to find out what is causing a huge slowdown, then optimize that portion.
+- Test if your improvement actually speeds up the runtime in practice.
+- If trying to optimize a certain portion to run slightly faster will take you a few days, avoid that. Find a balance between delivering the product or features, and scalability.
+
+## Conclusion
+
+There is a huge gap between theory and practice. What looks well on paper might need tweaking with arbitrary constants and "hacks". As such, more research should be done before you implement any of the solutions.
+
+However, scalability will still be a crucial part of servers aiming to provide services to an increasingly-growing user base. Learning good scalability practices will help prevent developing a server that will never be able to scale. Take the time to explore the depths of scalability, and you will be able help your server scale to meet its demand.
```

Comment:
Overall, the document seems to be a mix of a general introduction about scalability and specific tips aimed at improving a certain kind of applications. For example, the article seems to assume the software has a server backend, which is not actually a prerequisite to writing scalable code. Scalability is applicable to software other can client-server software.